doc/rm/asm_coding_style.md

```diff
@@ -0,0 +1,286 @@
+
+***CSRs defined in the RISC-V spec must be refered to by those names (like `mstatus`), while custom non-standard ones must be encapsulated in a `#define`.***
+
+Naturally, if a pseudoinstruction exists to read that CSR (like `rdtime`) that one should be used, instead.
+
+`#define`s for CSRs should be prefixed with `CSR_<design>_`, where `<design>` is the name of the design the CSR corresponds to.
+
+Recognized CSR prefixes:
+- `CSR_IBEX_` - A CSR specific to the Ibex core.
+- `CSR_OT_` - A CSR specific to the OpenTitan chip, beyond the Ibex core.
+
+Example:
+```S
+  csrr t0, mstatus
+
+  #define CSR_OT_HMAC_ENABLED ...
+  csrw CSR_OT_HMAC_ENABLED, 0x1
+```
+
+### Load and Store From Pointer in Register
+
+***When loading and storing from a pointer in a register, prefer to use `n(reg)` shorthand.***
+
+In the case that a pointer is being read without an offset, prefer `0(reg)` over `(reg)`.
+
+```S
+  // Correct:
+  lw t3, 8(sp)
+  sb t3, 0(a0)
+  // Wrong:
+  lw t3, sp, 8
+  sb t3, a0
+```
+
+### Compressed Instruction Mnemonics
+
+***Do not use compressed instruction mnemonics.***
+
+While Ibex implements the RISC-V C extension, it is expected that the toolchain will automatically compress instructions where possible.
+This advice extends to the `c.unimp` all-zeroes illegal instruction.
+Toolchains like `LLVM` and `gas` are expected to recognize the `unimp` mnemonic to stand for two `c.unimp` instructions.
+
+Of course, this advice should be ignored when it is necessary to prove that a certain block of instructions does not exceed a particular width.
+
+### "Current Point" Label
+
+***Do not use the current point (`.`) label.***
+
+The current point label does not look like a label, and can be easilly missed during review.
+
+## `.S` Files
+
+This advice applies specifically to `.S` files, as well as globally-scoped assembly in `.c` and `.cc` files.
+
+While this is is already implicit, we only use the `.S` extension for assembly files; not `.s` or `.asm`.
+
+### Indentation
+
+***Assembly files must be formatted with all directives indented two spaces, except for labels. Comments should be indented as usual.***
+
+Example:
+```S
+_trap_start:
+  .globl _trap_start
+  csrr a0, mcause
+  // ...
+```
+
+### Comments
+
+***Comments must use either the `//` or `/* */` syntaxes.***
+
+Every function-like label which is meant to be called like a function (*especially* `.globl`s) should be given a Doxygen-style comment.
+While Doxygen is not suited for assembly, that style should be used for consistency.
+See the [C/C++ style guide]({{< relref "c_cpp_coding_style" >}}) for more information.
+
+All other advice for writing comments, as in the C/C++ style guide, also applies.
+
+### Register useage
+
+***Register usage in a "function" that diverges from the RISC-V function call ABI must be documented.***
+
+This includes non-standard calling conventions, non-standard clobbers, and other behavior not expected of a well-behaved RISV-V function.
```

Comment:
Don't we want similar annotations for assembly functions that do match the RISC-V function call ABI too?

I think it would also be worth adding some guidance on annotating register usage within a function too. Maybe something like "Although lengthy assembly functions are undesirable, if they are necessary care should be taken to annotate the assignment of logical values to registers within the function in a way that makes it easier to understand."?

Reply:
Done.